once/__init__.py

```diff
@@ -25,110 +21,165 @@ def _is_method(func: collections.abc.Callable):
+                iterator = self._execute_call_once_async_iter(func, *args, **kwargs)
+                while True:
+                    try:
+                        next_value = yield await iterator.asend(next_value)
+                    except StopAsyncIteration:
+                        return
+
+        elif self.fn_type == _WrappedFunctionType.ASYNC_FUNCTION:
+
+            async def wrapped(*args, **kwargs):
+                return await self._execute_call_once_async(func, *args, **kwargs)
+
+        elif self.fn_type == _WrappedFunctionType.SYNC_FUNCTION:
+
+            def wrapped(*args, **kwargs):
+                return self._execute_call_once_sync(func, *args, **kwargs)
+
+        else:
+            if self.fn_type != _WrappedFunctionType.SYNC_GENERATOR:
+                print(self.fn_type)
+            assert self.fn_type == _WrappedFunctionType.SYNC_GENERATOR
+
+            def wrapped(*args, **kwargs):
+                yield from self._execute_call_once_sync_iter(func, *args, **kwargs)
+
+        functools.update_wrapper(wrapped, func)
+        return wrapped
 
     async def _execute_call_once_async(self, func: collections.abc.Callable, *args, **kwargs):
-        if self.called:
-            return self.return_value
         async with self.async_lock:
-            if self.called:
-                return self.return_value
-            else:
-                self.return_value = await func(*args, **kwargs)
-            self.called = True
-            return self.return_value
-
-    # This cannot be an async function!
-    def _execute_call_once_async_iter(self, func: collections.abc.Callable, *args, **kwargs):
-        if self.called:
-            return self.return_value.yield_results()
-        with self.lock:
-            if not self.called:
-                self.called = True
+            call_state = self.call_state
+        while call_state != _ExecutionState.COMPLETED:
+            if call_state == _ExecutionState.WAITING:
+                # Allow another thread to grab the GIL.
+                await asyncio.sleep(0)
+            async with self.async_lock:
+                call_state = self.call_state
+                if call_state == _ExecutionState.UNCALLED:
+                    self.call_state = _ExecutionState.WAITING
+            # Only one thread will be allowed into this state.
+            if call_state == _ExecutionState.UNCALLED:
+                try:
+                    return_value = await func(*args, **kwargs)
+                except Exception as exc:
+                    async with self.async_lock:
+                        self.call_state = _ExecutionState.UNCALLED
+                    raise exc
+                async with self.async_lock:
+                    self.return_value = return_value
+                    self.call_state = _ExecutionState.COMPLETED
+        return self.return_value
+
+    async def _execute_call_once_async_iter(self, func: collections.abc.Callable, *args, **kwargs):
```

Comment:
I'm sure this has something to do with generators, but why don't we have the WAITING/SLEEP logic here? And in the sync_iter?

Reply:
Wait this is handled for us in _iterator_wrappers right?

Reply:
Yes, exactly - these just instantiate the iterator wrapper, which does the rest of the work.